crossdock/server/oneway/echo.go

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package oneway
+
+import (
+	"bytes"
+	"context"
+	"time"
+
+	"go.uber.org/yarpc"
+	"go.uber.org/yarpc/encoding/raw"
+	"go.uber.org/yarpc/transport"
+	"go.uber.org/yarpc/transport/http"
+)
+
+const callBackAddr = "http://127.0.0.1:8089"
+
+// EchoRaw implements the echo/raw procedure.
+func EchoRaw(ctx context.Context, reqMeta yarpc.ReqMeta, body []byte) error {
+	//make call back to the client
+	callHome(body)
+	return nil
+}
+
+// EchoJSON implements the echo/json procedure.
+func EchoJSON(ctx context.Context, reqMeta yarpc.ReqMeta, body map[string]interface{}) error {
+	token := body["token"].(string)
+
+	//make call back to the client
+	callHome([]byte(token))
+	return nil
+}
+
+type onewayHandler struct{}
+
+// Echo implements the Oneway::Echo procedure.
+func (onewayHandler) Echo(ctx context.Context, reqMeta yarpc.ReqMeta, Token *string) error {
+	callHome([]byte(*Token))
+	return nil
+}
+
+func callHome(body []byte) {
+	// reduce the chance of a race condition
+	time.Sleep(time.Millisecond * 100)
+
+	outbound := http.NewOutbound(callBackAddr)
+	outbound.Start(transport.NoDeps)
```

Comment:
The outbound address is known in advance. You can avoid setting this up every
call. What you probably want to do here is,

```
type onewayHandler struct {
  Outbound transport.OnewayOutbound
}

func (o *onewayHandler) EchoJSON(...) { ... }
func (o *onewayHandler) EchoRaw(...) { ... }
func (o *onewayHandler) Echo(...) { ... }
```

And register these procedures against a bound instance of `onewayHandler`.

```
h := onewayHandler{Outbound: out}
dispatcher.Register(raw.OnewayProcedure(h.EchoRaw))
dispatcher.Register(json.OnewayProcedure(h.EchoJSON))
dispatcher.Register(onewayserver.New(h))
```

Although I don't feel strongly about this.